frankenphp.go

```diff
@@ -131,6 +131,22 @@ type FrankenPHPContext struct {
 
 	done                 chan interface{}
 	currentWorkerRequest cgo.Handle
+
+	hooks *FrankenPHPHooks
+}
+
+// ResponseFilter A callback that takes the current http status code, headers, and request. If it returns true, then
+// any further PHP output will be sent to stderr. This is called after all headers are sent and before a body is
+// written.
+type ResponseFilter func(status int, headers http.Header, r *http.Request) bool
+
+// FrankenPHPHooks You can pass this to FrankenPHP to "hook into" various processes and direct how FrankenPHP will
+// proceed. For example, hooking into responses and redirecting based on headers; detecting db credentials in output
+// and sending a white page instead; or, ensuring a custom error page is always shown even if it is a 500 error.
+type FrankenPHPHooks struct {
```

Comment:
To be candid, that seems like a pre-generalization (an architectural pre-optimization that might pigeonhole what things we can support in the future). As this is the only hook that we know of, we have no way to know what parameters can be generalized because other hooks would exist during different phases of the request/response cycle.

It's code, so we can technically do whatever we want with it, but I would advise waiting to generalize until there is a need for it. 

More technically, the hook is called after all headers have been written, but before a response code is set, which guarantees the underlying writer hasn't flushed the headers to the client yet (since, generally, the status code must be the first line/frame in an HTTP response). We can't know what optimizations the writer makes, but we can know the http spec and use it to our advantage. Maybe this is the wrong way to look at it, but it's the assumption I took with me when writing the code.